python/ql/src/experimental/semmle/python/frameworks/SqlAlchemy.qll

```diff
@@ -0,0 +1,97 @@
+/**
+ * Provides classes modeling security-relevant aspects of the 'SqlAlchemy' package.
+ * See https://pypi.org/project/SQLAlchemy/.
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.dataflow.new.TaintTracking
+private import semmle.python.ApiGraphs
+private import semmle.python.Concepts
+
+private module SqlAlchemy {
+  /**
+   * An instantization of a SqlAlchemy Session object.
+   * See https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session and
+   * https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.sessionmaker
+   */
+  private class SqlAlchemySessionInstance extends API::Node {
+    SqlAlchemySessionInstance() {
+      this in [
+          API::moduleImport("sqlalchemy.orm").getMember("Session").getReturn(),
+          API::moduleImport("sqlalchemy.orm").getMember("sessionmaker").getReturn().getReturn()
+        ]
+    }
+
+    override string toString() { result = "Use of SqlAlchemy Session instantization" }
+  }
+
+  /**
+   * An instantization of a SqlAlchemy Engine object.
+   * See https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine
+   */
+  private class SqlAlchemyEngineInstance extends API::Node {
+    SqlAlchemyEngineInstance() {
+      this = API::moduleImport("sqlalchemy").getMember("create_engine").getReturn()
+    }
+
+    override string toString() { result = "Use of SqlAlchemy create_engine member" }
+  }
+
+  /**
+   * An instantization of a SqlAlchemy Query object.
+   * See https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=query#sqlalchemy.orm.Query
+   */
+  private class SqlAlchemyQueryInstance extends API::Node {
+    SqlAlchemyQueryInstance() {
+      this = any(SqlAlchemySessionInstance sessionInstance).getMember("query").getReturn()
+    }
+
+    override string toString() { result = "Use of SqlAlchemy Session Query member" }
+  }
+
+  /**
+   * A call to `execute` meant to execute an SQL expression
+   * See the following links:
+   *   - https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=execute#sqlalchemy.engine.Connection.execute
+   *   - https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=execute#sqlalchemy.engine.Engine.execute
+   *   - https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=execute#sqlalchemy.orm.Session.execute
+   */
+  private class SqlAlchemyExecuteCall extends DataFlow::CallCfgNode, SqlExecution::Range {
+    SqlAlchemyExecuteCall() {
+      exists(SqlAlchemySessionInstance sessionInstance, SqlAlchemyEngineInstance engineInstance |
+        this = sessionInstance.getMember("execute").getACall() or
+        this = engineInstance.getMember("connect").getReturn().getMember("execute").getACall() or
+        this = engineInstance.getMember("begin").getReturn().getMember("execute").getACall()
+      )
+    }
+
+    override DataFlow::Node getSql() { result = this.getArg(0) }
```

Comment:
There was some discussion between @jorgectf and I on this predicate and whether we really should have the predicate be called `getSqlNode()`. Any feedback is appreciated on this.

Reply:
Yeah, the point is that, since the function returns a `DataFlow::Node`, perhaps the naming would be more specific.

Reply:
No, I don't think `getSqlNode` would be better. It might be argued that we should have called this `getSqlArg` instead, to match the naming from `RouteSetup::Range` (shown below). However, for this case were we don't have a `getSqlString` member-predicate, I don't think it's going to be the end of the world :wink: 

https://github.com/github/codeql/blob/5c2bf68a05a895cd86d3cf32b70e045ce64782dc/python/ql/src/semmle/python/Concepts.qll#L342-L346



On a low level, you also _need_ to use `getSql`, since that is what we called it in `SqlExecution::Range`"

https://github.com/github/codeql/blob/5c2bf68a05a895cd86d3cf32b70e045ce64782dc/python/ql/src/semmle/python/Concepts.qll#L291-L292

Reply:
I just wanted to say that conversations like this is extremely helpful for me as I write CodeQL queries. I was just working on a different query and referenced this for how I should name a method.